crates/neon-runtime/src/napi/bindings/mod.rs

```diff
@@ -0,0 +1,59 @@
+macro_rules! napi_bindings {
+    (extern "C" {
+        $(fn $name:ident($($param:ident: $ptype:ty$(,)?)*) -> $rtype:ty;)+
+    }) => {
+        pub(crate) struct Napi {
+            $(
+                $name: unsafe extern "C" fn(
+                    $($param: $ptype,)*
+                ) -> $rtype,
+            )*
+        }
+
+        pub(crate) unsafe fn load() {
+            let host = Library::this();
+            #[cfg(windows)]
+            let host = host.unwrap();
+
+            NAPI = Napi {
+                $(
+                    $name: *host.get(stringify!($name).as_bytes()).unwrap(),
+                )*
+            };
+        }
+
+        $(
+            pub(crate) unsafe fn $name($($param: $ptype,)*) -> $rtype {
+                (NAPI.$name)($($param,)*)
+            }
+        )*
+
+        fn panic_load<T>() -> T {
+            panic!("Must load N-API bindings")
+        }
+
+        static mut NAPI: Napi = {
+            $(
+                unsafe extern "C" fn $name($(_: $ptype,)*) -> $rtype {
+                    panic_load()
+                }
+            )*
+
+            Napi {
+                $(
+                    $name,
+                )*
+            }
+        };
+    };
+}
+
+mod types;
+mod functions;
+
+pub(crate) use functions::*;
+
+// FIXME: Make this safe!
+pub unsafe fn setup() {
```

Comment:
I don't think this _needs_ to be safe since we are the only ones who will ever call it, once during initialization. Is there a way that this would be called more than once?

Reply:
I think using the module in multiple contexts might.

Reply:
[`std::sync::Once`](https://doc.rust-lang.org/std/sync/struct.Once.html) is exactly what I needed. The description even includes:
> Useful for one-time initialization for **FFI** or related functionality.

```rust
static SETUP: Once = Once::new();

pub fn setup() {
    SETUP.call_once(|| unsafe {
        load();
    });
}
```

This is completely safe. It can be called multiple times from multiple threads. The closure will only execute once and if it's called concurrently, the other calls will *block* until it has completed.